File: emulinker/src/main/java/org/emulinker/kaillera/relay/V086Relay.java
package org.emulinker.kaillera.relay;

import com.google.common.flogger.FluentLogger;
import dagger.assisted.Assisted;
import dagger.assisted.AssistedFactory;
import dagger.assisted.AssistedInject;
import java.net.InetSocketAddress;
import java.nio.Buffer;
import java.nio.ByteBuffer;
import org.emulinker.kaillera.controller.messaging.*;
import org.emulinker.kaillera.controller.v086.protocol.*;
import org.emulinker.net.UDPRelay;
import org.emulinker.util.EmuUtil;

public final class V086Relay extends UDPRelay {
  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  private int lastServerMessageNumber = -1;
  private int lastClientMessageNumber = -1;

  @AssistedFactory
  public interface Factory {
    V086Relay create(int listenPort, InetSocketAddress serverSocketAddress);
  }

  @AssistedInject
  V086Relay(@Assisted int listenPort, @Assisted InetSocketAddress serverSocketAddress) {
    super(listenPort, serverSocketAddress);
  }

  @Override
  public String toString() {
    return "Kaillera client datagram relay version 0.86 on port " + super.getListenPort();
  }

  @Override
  protected ByteBuffer processClientToServer(
      ByteBuffer receiveBuffer, InetSocketAddress fromAddress, InetSocketAddress toAddress) {
    V086Bundle inBundle = null;

    logger.atFine().log("-> " + EmuUtil.dumpBuffer(receiveBuffer));

    try {
      // inBundle = V086Bundle.parse(receiveBuffer, lastClientMessageNumber);
      inBundle = V086Bundle.parse(receiveBuffer, -1);
    } catch (ParseException e) {
      receiveBuffer.rewind();
      logger.atWarning().withCause(e).log("Failed to parse: " + EmuUtil.dumpBuffer(receiveBuffer));
      return null;
    } catch (V086BundleFormatException e) {
      receiveBuffer.rewind();
      logger.atWarning().withCause(e).log(
          "Invalid message bundle format: " + EmuUtil.dumpBuffer(receiveBuffer));
      return null;
    } catch (MessageFormatException e) {
      receiveBuffer.rewind();
      logger.atWarning().withCause(e).log(
          "Invalid message format: " + EmuUtil.dumpBuffer(receiveBuffer));
      return null;
    }

    logger.atFine().log("-> " + inBundle);

    V086Message[] inMessages = inBundle.getMessages();
    for (int i = 0; i < inBundle.getNumMessages(); i++) {
      if (inMessages[i].messageNumber() > lastClientMessageNumber)
        lastClientMessageNumber = inMessages[i].messageNumber();
    }

    ByteBuffer sendBuffer = ByteBuffer.allocate(receiveBuffer.limit());
    receiveBuffer.rewind();
    sendBuffer.put(receiveBuffer);
    // Cast to avoid issue with java version mismatch: https://stackoverflow.com/a/61267496/2875073
    ((Buffer) sendBuffer).flip();

    return sendBuffer;
  }

  @Override
  protected ByteBuffer processServerToClient(
      ByteBuffer receiveBuffer, InetSocketAddress fromAddress, InetSocketAddress toAddress) {
    V086Bundle inBundle = null;

    logger.atFine().log("<- " + EmuUtil.dumpBuffer(receiveBuffer));

    try {
      // inBundle = V086Bundle.parse(receiveBuffer, lastServerMessageNumber);
      inBundle = V086Bundle.parse(receiveBuffer, -1);
    } catch (ParseException e) {
      receiveBuffer.rewind();
      logger.atWarning().withCause(e).log("Failed to parse: " + EmuUtil.dumpBuffer(receiveBuffer));
      return null;
    } catch (V086BundleFormatException e) {
      receiveBuffer.rewind();
      logger.atWarning().withCause(e).log(
          "Invalid message bundle format: " + EmuUtil.dumpBuffer(receiveBuffer));
      return null;
    } catch (MessageFormatException e) {
      receiveBuffer.rewind();
      logger.atWarning().withCause(e).log(
          "Invalid message format: " + EmuUtil.dumpBuffer(receiveBuffer));
      return null;
    }

    logger.atInfo().log("<- " + inBundle);

    V086Message[] inMessages = inBundle.getMessages();
    for (int i = 0; i < inBundle.getNumMessages(); i++) {
      if (inMessages[i].messageNumber() > lastServerMessageNumber)
        lastServerMessageNumber = inMessages[i].messageNumber();
    }

    ByteBuffer sendBuffer = ByteBuffer.allocate(receiveBuffer.limit());
    receiveBuffer.rewind();
    sendBuffer.put(receiveBuffer);
    // Cast to avoid issue with java version mismatch: https://stackoverflow.com/a/61267496/2875073
    ((Buffer) sendBuffer).flip();

    return sendBuffer;
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/relay/V086Relay.kt
package org.emulinker.kaillera.relay

import com.google.common.flogger.FluentLogger
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import java.net.InetSocketAddress
import java.nio.Buffer
import java.nio.ByteBuffer
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.messaging.ParseException
import org.emulinker.kaillera.controller.v086.protocol.V086Bundle
import org.emulinker.kaillera.controller.v086.protocol.V086Bundle.Companion.parse
import org.emulinker.kaillera.controller.v086.protocol.V086BundleFormatException
import org.emulinker.net.UDPRelay
import org.emulinker.util.EmuUtil.dumpBuffer

private val logger = FluentLogger.forEnclosingClass()

class V086Relay
    @AssistedInject
    internal constructor(
        @Assisted listenPort: Int, @Assisted serverSocketAddress: InetSocketAddress
    ) : UDPRelay(listenPort, serverSocketAddress) {
  private var lastServerMessageNumber = -1
  private var lastClientMessageNumber = -1

  @AssistedFactory
  interface Factory {
    fun create(listenPort: Int, serverSocketAddress: InetSocketAddress?): V086Relay?
  }

  override fun toString() =
      "Kaillera client datagram relay version 0.86 on port ${super.listenPort}"

  override fun processClientToServer(
      receiveBuffer: ByteBuffer, fromAddress: InetSocketAddress, toAddress: InetSocketAddress
  ): ByteBuffer? {
    logger.atFine().log("-> " + dumpBuffer(receiveBuffer))
    val inBundle: V086Bundle =
        try {
          // inBundle = V086Bundle.parse(receiveBuffer, lastClientMessageNumber);
          parse(receiveBuffer, -1)
        } catch (e: ParseException) {
          receiveBuffer.rewind()
          logger.atWarning().withCause(e).log("Failed to parse: " + dumpBuffer(receiveBuffer))
          return null
        } catch (e: V086BundleFormatException) {
          receiveBuffer.rewind()
          logger
              .atWarning()
              .withCause(e)
              .log("Invalid message bundle format: " + dumpBuffer(receiveBuffer))
          return null
        } catch (e: MessageFormatException) {
          receiveBuffer.rewind()
          logger
              .atWarning()
              .withCause(e)
              .log("Invalid message format: " + dumpBuffer(receiveBuffer))
          return null
        }
    logger.atFine().log("-> $inBundle")
    val inMessages = inBundle.messages
    for (i in 0 until inBundle.numMessages) {
      if (inMessages[i]!!.messageNumber > lastClientMessageNumber)
          lastClientMessageNumber = inMessages[i]!!.messageNumber
    }
    val sendBuffer = ByteBuffer.allocate(receiveBuffer.limit())
    receiveBuffer.rewind()
    sendBuffer.put(receiveBuffer)
    // Cast to avoid issue with java version mismatch: https://stackoverflow.com/a/61267496/2875073
    (sendBuffer as Buffer).flip()
    return sendBuffer
  }

  override fun processServerToClient(
      receiveBuffer: ByteBuffer, fromAddress: InetSocketAddress, toAddress: InetSocketAddress
  ): ByteBuffer? {
    logger.atFine().log("<- " + dumpBuffer(receiveBuffer))
    val inBundle: V086Bundle =
        try {
          // inBundle = V086Bundle.parse(receiveBuffer, lastServerMessageNumber);
          parse(receiveBuffer, -1)
        } catch (e: ParseException) {
          receiveBuffer.rewind()
          logger.atWarning().withCause(e).log("Failed to parse: " + dumpBuffer(receiveBuffer))
          return null
        } catch (e: V086BundleFormatException) {
          receiveBuffer.rewind()
          logger
              .atWarning()
              .withCause(e)
              .log("Invalid message bundle format: " + dumpBuffer(receiveBuffer))
          return null
        } catch (e: MessageFormatException) {
          receiveBuffer.rewind()
          logger
              .atWarning()
              .withCause(e)
              .log("Invalid message format: " + dumpBuffer(receiveBuffer))
          return null
        }
    logger.atInfo().log("<- $inBundle")
    val inMessages = inBundle.messages
    for (i in 0 until inBundle.numMessages) {
      if (inMessages[i]!!.messageNumber > lastServerMessageNumber)
          lastServerMessageNumber = inMessages[i]!!.messageNumber
    }
    val sendBuffer = ByteBuffer.allocate(receiveBuffer.limit())
    receiveBuffer.rewind()
    sendBuffer.put(receiveBuffer)
    // Cast to avoid issue with java version mismatch: https://stackoverflow.com/a/61267496/2875073
    (sendBuffer as Buffer).flip()
    return sendBuffer
  }
}
